### PR TITLE
fix: pipe output to files to avoid buffering and I/O ops, weird timeouts in plugin-packaging

### PR DESCRIPTION
**Pull request by WillieRuemmele**
[this](https://github.com/salesforcecli/plugin-packaging/pull/198/commits/e68a85e14431b7d1b462bd691729620d3760c6bf) commit in plugin-packaging started seeing weird timeouts like 
```
{
        "Id": "08c4p000000081bAAA",
        "Status": "Success",
        "Package2Id": "0Ho4p00000003x6CAA",
        "Package2VersionId": "05i4p00000007azAAA",
        "SubscriberPackageVersionId": "04t4p000001zyDGAAY",
        "Tag": null,
        "Branch": null,
        "Error": [],
        "CreatedDate": "2023-01-17 09:16",
        "HasMetadataRemoved": false,
        "CreatedBy": "00546000002jDFFAA2"
      },
      {
        "Id": "08c4p00000008
  stderr=Error: timed out
      at Object.error (node_modules/@oclif/core/lib/errors/index.js:28:15)
      at /Users/william.ruemmele/projects/oss/plugin-packaging/node_modules/@oclif/core/lib/cli-ux/index.js:25:66
      at async flush (node_modules/@oclif/core/lib/cli-ux/index.js:124:9)
  
      at getExitCodeError (/Users/william.ruemmele/projects/oss/cli-plugins-testkit/src/execCmd.ts:104:10)
      at execCmdSync (/Users/william.ruemmele/projects/oss/cli-plugins-testkit/src/execCmd.ts:170:11)
      at execCmd (/Users/william.ruemmele/projects/oss/cli-plugins-testkit/src/execCmd.ts:248:12)
      at Context.<anonymous> (test/commands/package/packageVersion.nut.ts:182:29)
      at processImmediate (node:internal/timers:471:21)
```
it was like it was timing out while printing to the `shelljs.exec` method, trying increasing the Buffer size and timeout had no affect. It was very weird because they were passing on windows, but the linux vm was almost too fast??

In an effort to reduce operations, Pete suggested piping the command output to a file, and then reading it back into memory. When implemented in the plugin this resolved the issue. Now, while linking this PR to that commit in PP, all the NUTs pass 8 mins, vs the 25 it took for all of them to fail.

We can move this into the NUTs library to help all NUTs 

plugin-user ✅ 
plugin-packaging ✅ 
plugin-config ✅ 
plugin-alias ✅ 
plugin-limits ✅ 
plugin-data ✅ 
plugin-source ✅ 

@W-12437558@

### CODE REVIEW

File: src/execCmd.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@
  */
 import * as fs from 'fs';
 import { spawn, SpawnOptionsWithoutStdio } from 'child_process';
-import { join as pathJoin, resolve as pathResolve } from 'path';
+import { join, join as pathJoin, resolve as pathResolve } from 'path';
```
Comment (peternhale) [inline src/execCmd.ts:9]: @WillieRuemmele join is already imported as `pathJoin`

File: src/execCmd.ts [hunk 2]
```diff
@@ -153,38 +153,50 @@ const execCmdSync = <T>(cmd: string, options?: ExecCmdOptions): ExecCmdResult<T>
...
+    shellOutput: new ShellString(fs.readFileSync(commandFileLocation, 'utf-8')),
     execCmdDuration: Duration.seconds(0),
   } as ExecCmdResult<T>;
 
   // Execute the command in a synchronous child process
   const startTime = process.hrtime();
-  result.shellOutput = shelljs.exec(cmd, cmdOptions) as ShellString;
-  result.shellOutput.stdout = stripAnsi(result.shellOutput.stdout);
-  result.shellOutput.stderr = stripAnsi(result.shellOutput.stderr);
+  result.shellOutput.code = (shelljs.exec(`${cmd} > ${commandFile} 2> ${stderrFile} `, cmdOptions) as ShellString).code;
```
Comment (peternhale) [inline src/execCmd.ts:168]:
``` suggestion
  result.shellOutput.code = (shelljs.exec(`${cmd} 1> ${commandFile} 2> ${stderrFile} `, cmdOptions) as ShellString).code;
```

File: src/execCmd.ts [hunk 3]
```diff
@@ -197,22 +209,18 @@ const execCmdAsync = async <T>(cmd: string, options: ExecCmdOptions): Promise<Ex
...
-      resolve(addJsonOutput<T>(cmd, result));
+      result.shellOutput.stdout = stripAnsi(fs.readFileSync(commandFilePath, 'utf-8'));
+      result.shellOutput.stderr = stripAnsi(fs.readFileSync(stderrFilePath, 'utf-8'));
+      resolve(addJsonOutput<T>(cmd, result, commandFilePath));
     };
-
     // Execute the command async in a child process
     const startTime = process.hrtime();
-    shelljs.exec(cmd, cmdOptions, callback);
+    shelljs.exec(`${cmd} > ${commandFile} 2> ${stderrFile}`, cmdOptions, callback);
```
Comment (peternhale) [inline src/execCmd.ts:222]:
``` suggestion
    shelljs.exec(`${cmd} 1> ${commandFile} 2> ${stderrFile}`, cmdOptions, callback);
```

**Review by peternhale — CHANGES_REQUESTED** [review]

File: src/execCmd.ts [hunk 4]
```diff
@@ -153,38 +153,50 @@ const execCmdSync = <T>(cmd: string, options?: ExecCmdOptions): ExecCmdResult<T>
...
 
   if (isNumber(cmdOptions.ensureExitCode) && result.shellOutput.code !== cmdOptions.ensureExitCode) {
     throw getExitCodeError(cmd, cmdOptions.ensureExitCode, result.shellOutput);
   }
 
-  return addJsonOutput<T>(cmd, result);
+  return addJsonOutput<T>(cmd, result, commandFileLocation);
 };
 
+const genFileName = (name: string): string => `${name}-${Math.random()}.txt`;
```
Comment (shetzel) [inline src/execCmd.ts:185]: You could instead use `genUniqueString()` function from this lib.

File: test/unit/execCmd.test.ts [hunk 5]
```diff
@@ -33,7 +40,9 @@ describe('execCmd (sync)', () => {
     const shellString = new ShellString(JSON.stringify(output));
     const execStub = stubMethod(sandbox, shelljs, 'exec').returns(shellString);
     execCmd(cmd);
-    expect(execStub.args[0][0]).to.equal(`${binPath} ${cmd}`);
+    expect(execStub.args[0][0]).to.include(`${binPath} ${cmd}`);
+    expect(execStub.args[0][0]).to.include('> command');
```
Comment (peternhale) [inline test/unit/execCmd.test.ts:44]: @WillieRuemmele I know this works, but should these be changed to `1> command` for completeness?

**Review by peternhale — CHANGES_REQUESTED** [review]

**Review by shetzel — APPROVED** [review]

**Pull request merged**